### ISSUE TITLE
TcpListener.next() receives too many open files error indefinitely on Linux

### ISSUE THREAD

**Issue by himajin2045** [tokio-rs/tokio#3529]
**Version**
testtokio v0.1.0 (/root/rust/testtokio)
└── tokio v0.2.25
    └── tokio-macros v0.2.6

**Platform**
Linux vps 4.9.0-8-amd64 #1 SMP Debian 4.9.110-3+deb9u6 (2018-10-08) x86_64 GNU/Linux

**Description**
The program outputs `too many open files` errors indefinitely once hit the error.

I tried this code:

```rust
use futures::stream::Stream;
use std::net::SocketAddr;
use tokio::io::AsyncReadExt;
use tokio::net::{TcpListener, TcpStream};
use tokio::stream::StreamExt;

async fn process_stream(mut s: TcpStream) {
    let mut buf = vec![0u8; 2 * 1024];
    s.read_exact(&mut buf).await;
}

#[tokio::main]
async fn main() {
    let mut listener = TcpListener::bind("127.0.0.1:8080").await.unwrap();
    println!("listening");
    while let Some(stream) = listener.next().await {
        match stream {
            Ok(stream) => {
                println!("new client!");
                tokio::spawn(process_stream(stream));
            }
            Err(e) => {
                println!("accept err: {}", e);
            }
        }
    }
}
```

And in order to reproduce the `too many open files` error, limit the current session open files to 10:
```sh
ulimit -n 10
```

Then run `nc` multiple times to open some connections to `127.0.0.1:8080` and make sure we hit the `too many open files` error:
```sh
nc 127.0.0.1 8080
```

Then once the `too  many open files` error  occur, it will never  stop.

I tested the same scenario on macOS and it's OK, only reproducible on Linux boxes.

**Comment by Darksonn** [comment]
What would you expect it to do instead?

**Comment by himajin2045** [comment]
I expect it would receive only 1 error per accepted connection in case the available fds are exhausted. And new connections can be accepted once there are available fds, e.g. some previous connections closed.

**Comment by Darksonn** [comment]
I may be wrong, but I don't think it is possible for us to register for a wakeups when a file descriptor becomes available.

**Comment by himajin2045** [comment]
So I'm doing it the wrong way?

Is this the wrong way to accept tcp conns?

```rust
    while let Some(stream) = listener.next().await {
    }
```

Or in case we hit the error, we should do something else, e.g. exit the loop?

```rust
    while let Some(stream) = listener.next().await {
        match stream {
            Ok(stream) => {
                println!("new client!");
                tokio::spawn(process_stream(stream));
            }
            Err(e) => {
                break;
            }
        }
    }
```

**Comment by himajin2045** [comment]
I'm not sure if you get me right, I mean, it's an infinite loop. I limit the fds to 9 (`ulimit -n 9`), and run `nc` **only twice**, 2 connections, then I receive many errors, never ends:
```
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
accept err: Too many open files (os error 24)
...
...
...
```

**Comment by Darksonn** [comment]
I mean, I understand the problem. It's just that I don't know of any way to avoid it other than having some other mechanism of counting the number of connections. I don't think it is possible for Tokio to wait for file descriptors to become available. 

You should probably try an approach like creating a [`Semaphore`][1], putting it in an `Arc`, using `acquire_owned` before each accept, and moving the acquired permit into the new task, finally dropping the permit only after closing the fd.

[1]: https://docs.rs/tokio/1/tokio/sync/struct.Semaphore.html

**Comment by himajin2045** [comment]
Thanks for explaining this. I don't known any low-level details of  this issue, I assume it's not a bug in tokio so I'm going to close the issue.
